### ISSUE TITLE
Defragmentation of the Flash memory map

### ISSUE THREAD

**Issue by rogerclarkmelbourne** [rogerclarkmelbourne/Radioddity_GD-77#26]
The external SPI Flash memory is very fragmented and wastefully used.

In order for more of it to be used for the DMR ID some radical changes need to be made 

The region of flash from 0x10000 to 0x1FFFF , looks like it’s full of junk as does the region from 0x20000 to 0x2FFFF.

I backed up my entire flash memory and then wrote back zeros into those regions, and as far as I can tell, the radio still worked fine.

I will get a few more people to test and confirm this..

I also changed the code which calculates the address of the part of the codeplug in flash.

But I think I must not have changed everything that was necessary, because it did do as I expected.

There is code which adds 0x70000 to the address, so that when the CPS writes to the upper half of the codeplug at addresses 0x10000 and above, the add gets changed to 0x80000 

There only seemed to be one place I needed to patch, and change 0x70000 to 0, but something went wrong and the CPS didn’t seem to write to 0x10000

I probably need to check whether Ghidra patched instruction was correct, as it gets 4 byte instructions wrong sometimes.

I also used FlashManager to write zeros into the region that is currently the top half of the codeplug, at 0x80000, but I the green LED illuminated, and would not go out.

So I decided to revert both the firmware and restore the flash and things went back to normal.

BTW.
Jason, said that he seems to have somehow blocked all access to his Flash, when messing around with the region that holds the bitmap character data.

Apparently FlashManager can’t access it at all, but I’m not sure if the CPS can access the flash.

I suspect the code is stuck in a loop at boot up while trying to load bitmap or other data, which has frozen one of the RTOS tasks

**Comment by forkoz** [comment]
I see 0x7000 in the following palces:

split_codeplug_26436, in 2 spots. 
result += *a4 + 0x70000;
 result += 0x70000;

00026578 spi_read_some_cal_data   << hit by cps commands & fill flash w/ zeros
 if ( address + 0x70000 < 0x8F000 && address >= 0xB000 )
    address = spi_read_to_struct(0x20, address + 0x70000, a3, semaphore);

 Calibration_backup_SPI_Flash_0002616c. which theoretically shouldn't run.

**Comment by rogerclarkmelbourne** [comment]
yep. I knew calibration would be affected. That would need another patch

**Comment by forkoz** [comment]
You can't add 0, THe first address of the CP is B1b0 + 7K = 7B1B0

**Comment by forkoz** [comment]
**Lets do this then:**
### 1
```
ROM:0002644A 010                 STRH    R1, [R5]        ; Store to Memory
ROM:0002644C 010                 ADD.W   R0, R0, #0x10000 ; Rd = Op1 + Op2
ROM:00026450 010                 STR     R0, [R4]        ; Store to Memory
ROM:00026452 010                 B       loc_2647C       ; Branch
```
### 2
```
ROM:00026470 010                 LDRH    R1, [R3]        ; Load from Memory
ROM:00026472 010                 UXTAH.W R0, R0, R1      ; Unsigned extend halfword to word, add
ROM:00026476 010                 ADD.W   R0, R0, #0x10000 ; Rd = Op1 + Op2
ROM:0002647A 010                 STR     R0, [R4]        ; Store to Memory
```

And it works :)


I dunno what this function is used for:
26578     spi_read_some_cal_data  
```
unsigned int __fastcall spi_read_some_cal_data(unsigned int address, int a2, int a3, signed int semaphore)
{
  if ( address + 0x70000 < 0x8F000 && address >= 0xB000 )
    address = spi_read_to_struct(0x20, address + 0x70000, a3, semaphore);
  return address;
}
```

I didn't change it but I can read and write with CPS just fine.

**Comment by forkoz** [comment]
**Now we move calibration data!**
### 1
```
ROM:0002693C 000                 ADD.W   R0, R0, #0x2F000 ; Rd = Op1 + Op2
ROM:00026940 000                 CMP.W   R0, #0x30000    ; Set cond. codes on Op1 - Op2
ROM:00026944 000                 BCS     locret_26960    ; Branch
ROM:00026946 000                 UXTAH.W R3, R0, R2      ; Unsigned extend halfword to word, add
ROM:0002694A 000                 CMP.W   R3, #0x30000    ; Set cond. codes on Op1 - Op2
ROM:0002694E 000                 BCC     loc_26952       ; Branch
ROM:00026950 000                 NEGS    R2, R0          ; Negate
```

### 2

```
ROM:00026962 000                 ADD.W   R0, R0, #0x2F000 ; Rd = Op1 + Op2
ROM:00026966 000                 CMP.W   R0, #0x30000    ; Set cond. codes on Op1 - Op2
ROM:0002696A 000                 BCS     locret_2697E    ; Branch
ROM:0002696C 000                 UXTAH.W R3, R0, R2      ; Unsigned extend halfword to word, add
ROM:00026970 000                 CMP.W   R3, #0x30000    ; Set cond. codes on Op1 - Op2
ROM:00026974 000                 BCC     loc_26978       ; Branch
ROM:00026976 000                 NEGS    R2, R0          ; Negate
```

Then I move calibration data and have 0x30000 - 0xD0000 free. Everything seems to work ok.

**Comment by forkoz** [comment]
**Playing the dangerous game**

Patch
`0001F9BC 028                 B       allaccess_1F9EE ; Branch`

I have overwritten 
0xD0000 & 0xD0001 (values 80 & 00 respectively)
0xD0B30 & 0xD0B31 (values D2 & 84 respectively)

with FF's now, still have access.

**Comment by forkoz** [comment]
[318-DangerousGame.zip](https://github.com/rogerclarkmelbourne/Radioddity_GD-77/files/3101565/318-DangerousGame.zip)

If you want to try it.

**Comment by rogerclarkmelbourne** [comment]
How did you move all calibration data etc to a new location.
Did you read it with FlashManager then cut and paste to a new location using HxD and then write it to that new location using FlashManager

The problem with such a major change is that I suspect hardly anyone will bother to use it, as the majority of people would not want to use FlashManager and an editor etc etc before upgrading.


BTW. We need to prevent the situation where the user can be locked out if the Flash, like Jason was.

I will need to review what Jason told me about this and try to work out why the code prevents full access to the flash if some bytes are cleared.

**Comment by forkoz** [comment]
Simple, the code limits you to 

```
        contents = &contents_f_read;
        start_address_f_read = 0xD0000;         // WUT?
        end_address_f_read = 0xFFFFF;
        return contents;
```

If those bytes don't match.

**Comment by forkoz** [comment]
FWIW, we could take over one of the "soft reset" functions to:
* Check if "A0" is at 0x2F000
* If not, copy calibration from 0x8F000
* Copy font to 0xFE000
A "run once" upgrader so people don't have to hex edit so much.
